File: zig/2024/06/main.zig
pub fn main() !void {
    // var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    // defer _ = gpa.deinit();
    // const alloc = gpa.allocator();

    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    var args = try std.process.ArgIterator.initWithAllocator(alloc);
    defer args.deinit();

    _ = args.next();
    const filename = args.next().?;

    var p = try Puzzle.init(alloc, filename);
    defer p.deinit();

    try p.solve();
}

fn readFile(alloc: Allocator, filename: []const u8) ![]u8 {
    const file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    const file_size = try file.getEndPos();
    const data = try alloc.alloc(u8, file_size);

    _ = try file.readAll(data);
    return data;
}

const Vec = @Vector(2, i16);
const Guard = struct { pos: Vec, dir: u2 = 0 };
const Walker = struct {
    guard: Guard,
    walls: AutoHashMap(Vec, void),
    size: Vec,

    const Self = @This();
    const dirs = [4]Vec{ .{ 0, -1 }, .{ 1, 0 }, .{ 0, 1 }, .{ -1, 0 } };

    pub fn init(p: *const Puzzle) !Self {
        return Self{
            .guard = .{ .pos = p.start },
            .walls = try p.walls.clone(),
            .size = p.size,
        };
    }

    pub fn deinit(self: *Self) void {
        self.walls.deinit();
    }

    pub fn next(self: *Self) void {
        const p = self.guard.pos;
        var out: Vec = p + dirs[self.guard.dir];
        while (self.walls.contains(out)) {
            self.guard.dir +%= 1;
            out = p + dirs[self.guard.dir];
        }
        self.guard.pos = out;
    }

    pub fn inBounds(self: Self) bool {
        const x, const y = self.guard.pos;
        const w, const h = self.size;
        return (x >= 0 and y >= 0 and x < w and y < h);
    }
};

const Puzzle = struct {
    alloc: Allocator,
    walls: AutoHashMap(Vec, void),
    start: Vec,
    size: Vec,

    const Self = @This();

    pub fn init(alloc: Allocator, filename: []const u8) !Self {
        var walls = AutoHashMap(Vec, void).init(alloc);
        var s: ?Vec = null;
        var x: i16 = 0;
        var y: i16 = 0;

        const data = try readFile(alloc, filename);
        defer alloc.free(data);

        var lines = std.mem.tokenizeAny(u8, data, " \r\n");
        while (lines.next()) |l| {
            x = 0;
            for (l) |ch| {
                const v = .{ x, y };
                switch (ch) {
                    '#' => try walls.put(v, {}),
                    '^' => s = v,
                    else => {},
                }
                x += 1;
            }
            y += 1;
        }

        return .{
            .alloc = alloc,
            .walls = walls,
            .start = s.?,
            .size = .{ x, y },
        };
    }

    pub fn deinit(self: *Self) void {
        self.walls.deinit();
    }

    pub fn solve(self: *Self) !void {
        var visited = AutoHashMap(Vec, void).init(self.alloc);
        defer visited.deinit();

        var w = try Walker.init(self);
        defer w.deinit();

        while (w.inBounds()) {
            try visited.put(w.guard.pos, {});
            w.next();
        }
        print("Part 1: {d}\n", .{visited.count()});

        const start: [2]i16 = self.start;

        var loop_count: i16 = 0;
        var spots = visited.keyIterator();
        while (spots.next()) |p| {
            const s = p.*;
            if ((s[0] != start[0] or s[1] != start[1]) and try self.willLoop(s)) {
                loop_count += 1;
            }
        }

        print("Part 2: {d}\n", .{loop_count});
    }

    pub fn willLoop(self: *Self, obj: Vec) !bool {
        var w = try Walker.init(self);
        defer w.deinit();

        try w.walls.put(obj, {});

        var past = AutoHashMap(Guard, void).init(self.alloc);
        defer past.deinit();

        while (w.inBounds()) {
            const res = try past.getOrPut(w.guard);
            if (res.found_existing) return true;
            _ = w.next();
        }
        return false;
    }
};

const std = @import("std");
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const AutoHashMap = std.AutoHashMap;
const print = std.debug.print;

File: go/2024/06/main.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
)

var directions = []Vec{
	{0, -1}, {1, 0}, {0, 1}, {-1, 0},
}

type Vec struct {
	x, y int
}

func (v Vec) Add(other Vec) Vec {
	return Vec{v.x + other.x, v.y + other.y}
}

type Guard struct {
	pos Vec
	dir int
}

type Puzzle struct {
	startpos Vec
	w, h     int
	walls    map[Vec]bool
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func readFile(filename string) (lines []string) {
	f, err := os.Open(filename)
	check(err)
	defer f.Close()
	s := bufio.NewScanner(f)
	for s.Scan() {
		lines = append(lines, s.Text())
	}
	check(s.Err())
	return lines
}

func load(filename string) (puzzle Puzzle) {
	puzzle.walls = make(map[Vec]bool)
	data := readFile(filename)
	for y, line := range data {
		for x, r := range line {
			switch r {
			case '#':
				puzzle.walls[Vec{x, y}] = true
			case '^':
				puzzle.startpos = Vec{x, y}
			}
		}
	}
	puzzle.h = len(data)
	puzzle.w = len(data[0])
	return puzzle
}

func main() {
	flag.Parse()
	puzzle := load(flag.Arg(0))

	trail := puzzle.PartOne()
	puzzle.PartTwo(trail)
}

func (p Puzzle) PartOne() map[Vec]bool {
	pos := p.startpos
	d := 0
	trail := make(map[Vec]bool)
	for pos.x >= 0 && pos.y >= 0 && pos.x < p.w && pos.y < p.h {
		trail[pos] = true
		next := pos.Add(directions[d])
		if p.walls[next] {
			d = (d + 1) % 4
		} else {
			pos = next
		}
	}
	fmt.Println(len(trail))
	return trail
}

func (p Puzzle) PartTwo(trail map[Vec]bool) {
	count := 0
	for v := range trail {
		if v != p.startpos && p.WillLoop(v) {
			count++
		}
	}
	fmt.Println(count)
}

func (p Puzzle) WillLoop(block Vec) bool {
	walls := make(map[Vec]bool, len(p.walls))
	for v := range p.walls {
		walls[v] = true
	}
	walls[block] = true

	been := make(map[Guard]struct{})
	g := Guard{p.startpos, 0}
	for g.pos.x >= 0 && g.pos.y >= 0 && g.pos.x < p.w && g.pos.y < p.h {
		been[g] = struct{}{}
		pos := g.pos.Add(directions[g.dir])
		for walls[pos] {
			g.dir = (g.dir + 1) % 4
			pos = g.pos.Add(directions[g.dir])
		}
		g.pos = pos

		if _, ok := been[g]; ok {
			return true
		}
	}
	return false
}
